tooling/hibernate-ant/src/main/java/org/hibernate/tool/enhance/EnhancementTask.java

```diff
@@ -111,23 +126,44 @@ public void execute() throws BuildException {
 			return;
 		}
 
-		if ( !dir.startsWith( base ) ) {
-			throw new BuildException( "The enhancement directory 'dir' (" + dir + ") is no subdirectory of 'base' (" + base + ")" );
+		if ( base == null ) {
+			throw new BuildException( "The enhancement directory 'base' should be present" );
 		}
 
-		// Perform a depth first search for sourceSet
-		File root = new File( dir );
-		if ( !root.exists() ) {
-			log( "Skipping Hibernate enhancement task execution since there is no classes dir " + dir, Project.MSG_INFO );
-			return;
+		if ( !filesets.isEmpty() && dir != null ) {
+			throw new BuildException( "Please remove the enhancement directory 'dir' if 'fileset' is using" );
 		}
-		walkDir( root );
-		if ( sourceSet.isEmpty() ) {
-			log( "Skipping Hibernate enhancement task execution since there are no classes to enhance on " + dir, Project.MSG_INFO );
-			return;
+
+		for ( FileSet fileSet : filesets ) {
+			Iterator<Resource> it = fileSet.iterator();
+			while ( it.hasNext() ) {
+				File file = new File( it.next().toString() );
+				if ( file.isFile() ) {
+					sourceSet.add( file );
+				}
+			}
+		}
+
+		if ( dir != null ) {
+			if ( !dir.startsWith( base ) ) {
+				throw new BuildException( "The enhancement directory 'dir' (" + dir + ") is no subdirectory of 'base' (" + base + ")" );
+			}
+
+			// Perform a depth first search for sourceSet
+			File root = new File( dir );
+			if ( !root.exists() ) {
+				log( "Skipping Hibernate enhancement task execution since there is no classes dir " + dir, Project.MSG_INFO );
+				return;
+			}
+			walkDir( root );
+			if ( sourceSet.isEmpty() ) {
+				log( "Skipping Hibernate enhancement task execution since there are no classes to enhance on " + dir, Project.MSG_INFO );
+				return;
+			}
+
+			log( "Starting Hibernate enhancement task for classes on " + dir, Project.MSG_INFO );
 		}
```

Comment:
```suggestion
		if ( dir == null ) {
			for ( FileSet fileSet : filesets ) {
				Iterator<Resource> it = fileSet.iterator();
				while ( it.hasNext() ) {
					File file = new File( it.next().toString() );
					if ( file.isFile() ) {
						sourceSet.add( file );
					}
				}
			}
			
			if ( sourceSet.isEmpty() ) {
				log( "Skipping Hibernate enhancement task execution since there are no classes to enhance in the filesets " + filesets, Project.MSG_INFO );
				return;
			}
		}
		else {
			if ( !dir.startsWith( base ) ) {
				throw new BuildException( "The enhancement directory 'dir' (" + dir + ") is no subdirectory of 'base' (" + base + ")" );
			}

			// Perform a depth first search for sourceSet
			File root = new File( dir );
			if ( !root.exists() ) {
				log( "Skipping Hibernate enhancement task execution since there is no classes dir " + dir, Project.MSG_INFO );
				return;
			}
			walkDir( root );
			if ( sourceSet.isEmpty() ) {
				log( "Skipping Hibernate enhancement task execution since there are no classes to enhance on " + dir, Project.MSG_INFO );
				return;
			}

			log( "Starting Hibernate enhancement task for classes on " + dir, Project.MSG_INFO );
		}
```

Reply:
done.